### PR TITLE
mlir/Presburger: reinstate use of LogicalResult

### PR DESCRIPTION
**Pull request by artagnon**
Follow up on a desire post-landing d0fee98 (mlir/Presburger: strip dependency on MLIRSupport) to reinstate the use of LogicalResult in Presburger. Since db791b2 (mlir/LogicalResult: move into llvm), LogicalResult is in LLVM, and fulfilling this desire is possible while still maintaining the goal of stripping the Presburger library of mlir dependencies.

### CODE REVIEW

**Review by superty — APPROVED** [review]
Thanks for taking care of this! Please make sure the part I pointed out is correct before landing

File: mlir/lib/Analysis/Presburger/IntegerRelation.cpp [hunk 1]
```diff
@@ -2201,13 +2203,13 @@ bool IntegerRelation::unionBoundingBox(const IntegerRelation &otherCst) {
...
       // TODO: handle union if a dimension is unbounded.
-      return false;
+      return failure();
 
     auto otherExtent = otherCst.getConstantBoundOnDimSize(
         d, &otherLb, &otherLbFloorDivisor, &otherUb);
     if (!otherExtent.has_value() || lbFloorDivisor != otherLbFloorDivisor)
       // TODO: symbolic extents when necessary.
-      return false;
+      return success();
```
[the remark this answers is not in the document]
Comment (artagnon) [reply mlir/lib/Analysis/Presburger/IntegerRelation.cpp]: Thanks for catching this serious error not covered by tests. It was originally failure().

[the remark this answers is not in the document]
Comment (joker-eph) [reply mlir/lib/Analysis/Presburger/IntegerRelation.cpp]: Can you also add a test for this?

[the remark this answers is not in the document]
Comment (artagnon) [reply mlir/lib/Analysis/Presburger/IntegerRelation.cpp]: Will try to get to it in a follow-up patch.

**Pull request merged**